source/fab/tools/tool_box.py

```diff
@@ -0,0 +1,53 @@
+##############################################################################
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+##############################################################################
+
+'''This file contains the ToolBox class.
+'''
+
+from fab.tools.categories import Categories
+from fab.tools.tool import Tool
+from fab.tools.tool_repository import ToolRepository
+
+
+class ToolBox:
+    '''This class implements the tool box. It stores one tool for each
+    category to be used in a FAB build.
+    '''
+
+    def __init__(self):
+        self._all_tools = {}
```

Comment:
Good point, I still need to get my ahead around typing. BUT, when I also add the return type on `__getitem__`:
```
    def __getitem__(self, category: Categories) -> Tool:
```
I suddenly get 8 errors (that are no problem without this additional and unnecessary?? type information):
```
source/fab/steps/preprocess.py:158: error: Argument "preprocessor" to "pre_processor" has incompatible type "Tool"; expected "Preprocessor"  [arg-type]
```
All of them are kind of right, it gets a `Tool`, but expects a `Preprocessor`. What should I do?
1. Remove the return type of `__getitem`?
2. Use a type case everywhere to fix these errors?
3. Change the specific tool type in each case from (say) Preprocessor to Tool (though I expect I then get problems with  calling pre-processor specific functions in a 'tool'???  And I gave each tool a different `do_something` method (e.g. compile, link), since pylint does not like to overwrite a function (`run`) with one that takes different parameters - plus it's a lot more descriptive what is happening).

:(

Reply:
If a routine needs specifically a `Preprocessor` rather than a generic `Tool` then it should specify that in the argument list. It then becomes the caller's responsibility to make sure it is passing one. e.g.

```
if not isinstance(tool, Preprocessor):
    raise TypeError("Need a preprocessor")
thing.method(tool)
```

Alternatively you could do the check in the function

```
def method(self, tool: Tool):
    assert isinstance(tool, Preprocessor)
```

The first has the advantage of providing more static type checking but it is a bit ugly. The second is cleaner but reduces static type checking.

Reply:
Ah, that is great, I had no idea that mypy was that smart. I've done it the first way, if we go through all the effort of adding types, we should use them as much as possible.